### PR TITLE
Enhancement: Performance Improvements to blackman, hanning and hamming methods

### PR DESCRIPTION
**Pull request by niteya-shah**
This PR adds performance improvements to the following methods

- cupy.blackman
- cupy.hamming
- cupy.hanning

by incorporating Elementwise kernels.

The benchmarks are as follows

for cupy.blackman
size |  numpy |  This PR |  existing |
----- | ------ | ------ | ------ |
10 | 0.03 | 0.03 | 0.39 | 
100 | 0.02 | 0.03 | 0.39 | 
1000 | 0.05 | 0.03 | 0.39 | 
10000 | 0.35 | 0.03 | 0.39 | 
100000 | 4.12 | 0.05 | 0.50 | 
1000000 | 35.42 | 0.16 | 3.25 | 
10000000 | 403.96 | 1.33 | 22.36 | 

for cupy.hamming
size |  numpy |  This PR |  existing |
----- | ------ | ------ | ------ |
10 | 0.04 | 0.04 | 0.44 | 
100 | 0.03 | 0.07 | 0.43 | 
1000 | 0.05 | 0.03 | 0.24 | 
10000 | 0.41 | 0.04 | 0.24 | 
100000 | 3.36 | 0.04 | 0.36 | 
1000000 | 28.83 | 0.15 | 1.66 | 
10000000 | 327.42 | 1.18 | 11.22 | 

for cupy.hanning
size |  numpy |  This PR |  existing |
----- | ------ | ------ | ------ |
10 | 0.14 | 0.03 | 0.20 | 
100 | 0.02 | 0.04 | 0.39 | 
1000 | 0.08 | 0.05 | 0.33 | 
10000 | 0.43 | 0.04 | 0.23 | 
100000 | 3.47 | 0.06 | 0.27 | 
1000000 | 30.60 | 0.15 | 1.64 | 
10000000 | 328.86 | 1.24 | 11.25 |

### CODE REVIEW

**Comment by emcastillo** [comment]
Have you tried using `@cupy.fuse` instead of defining kernels by hand?
It may yield the same performance and be easier to maintain

**Comment by niteya-shah** [comment]
Thank you @emcastillo for the recommendation. I decorated the existing functions using  `@cupy.fuse` like
```python
@cupy.fuse()
def blackman_fuse(M):
    if M < 1:
        return from_data.array([])
    if M == 1:
        return basic.ones(1, float)
    n = ranges.arange(0, M)
    return 0.42 - 0.5 * trigonometric.cos(2.0 * numpy.pi * n / (M - 1))\
        + 0.08 * trigonometric.cos(4.0 * numpy.pi * n / (M - 1))
```
and got the following results,

for cupy.hamming
size |  numpy |  This PR |  existing | using fuse |
------ | ------- | -------- | -------- | --------- |
10 | 0.01 | 0.03 | 0.18 | 0.20 | 
100 | 0.01 | 0.03 | 0.18 | 0.19 | 
1000 | 0.02 | 0.03 | 0.18 | 0.33 | 
10000 | 0.14 | 0.03 | 0.20 | 0.19 | 
100000 | 1.26 | 0.03 | 0.24 | 0.24 | 
1000000 | 17.38 | 0.17 | 1.77 | 1.81 | 
10000000 | 165.29 | 1.19 | 11.22 | 11.17 | 

for cupy.blackman
size |  numpy |  This PR |  existing | using fuse |
------ | ------- | -------- | -------- | --------- |
10 | 0.03 | 0.03 | 0.65 | 0.39 | 
100 | 0.02 | 0.03 | 0.40 | 0.47 | 
1000 | 0.05 | 0.03 | 0.37 | 0.35 | 
10000 | 0.34 | 0.03 | 0.34 | 0.34 | 
100000 | 3.49 | 0.04 | 0.45 | 0.46 | 
1000000 | 28.24 | 0.15 | 3.17 | 3.27 | 
10000000 | 329.21 | 1.32 | 22.23 | 22.30 | 

for cupy.hanning
size |  numpy |  This PR |  existing | using fuse |
------ | ------- | -------- | -------- | --------- |
10 | 0.03 | 0.06 | 0.24 | 0.24 | 
100 | 0.01 | 0.03 | 0.25 | 0.40 | 
1000 | 0.03 | 0.03 | 0.29 | 0.28 | 
10000 | 0.18 | 0.04 | 0.23 | 0.22 | 
100000 | 1.58 | 0.06 | 0.26 | 0.26 | 
1000000 | 14.60 | 0.15 | 1.66 | 1.63 | 
10000000 | 170.99 | 1.18 | 12.83 | 11.18 | 

I seem to get very similar results to the original implementation, suggesting that kernel fusion cannot yet optimise this perfectly. Also a large improvement of my implementation comes from the fact that using elementwise kernels allows me to not call arrange and the necessarry range is produced automatically.

**Comment by emcastillo** [comment]
Thanks for trying it!

**Comment by takagi** [comment]
Would you try the following? I guess it would be slower than the kernel of this PR as `cupy.fuse` does not expose the element index `i` and `_blackman_kernel_fuse` requires the index range `n` instead.

```Python
@cupy.fuse()
def _blackman_kernel_fuse(alpha, n):
    return 0.42 - 0.5 * trigonometric.cos(n * alpha) + 0.08 * trigonometric.cos(2\
.0 * alpha * n)

def blackman_fuse(M):
    if M < 1:
	return from_data.array([])
    if M == 1:
	return basic.ones(1, float)
    alpha = (numpy.pi * 2)/(M - 1)
    n = ranges.arange(0, M)
    return _blackman_kernel_fuse(alpha, n)
```

**Comment by niteya-shah** [comment]
I ran the fused kernel you provided and got the following results
size = 1
```
blackman_fuse       :    CPU:   13.983 us   +/- 6.731 (min:   10.770 / max:  136.161) us     GPU:   16.995 us   +/- 7.643 (min:   11.392 / max:  147.104) us
blackman_PR         :    CPU:   11.998 us   +/- 4.128 (min:   10.760 / max:   84.762) us     GPU:   14.583 us   +/- 4.678 (min:   11.648 / max:   99.904) us
blackman            :    CPU:   12.907 us   +/- 6.123 (min:   10.752 / max:  194.381) us     GPU:   15.678 us   +/- 6.985 (min:   11.680 / max:  202.848) us
```

size = 10
```
blackman_fuse       :    CPU:   40.102 us   +/-13.505 (min:   34.225 / max:  458.219) us     GPU:   43.174 us   +/-14.297 (min:   35.296 / max:  466.624) us
blackman_PR         :    CPU:   12.825 us   +/- 6.050 (min:   10.857 / max:  231.848) us     GPU:   15.532 us   +/- 6.775 (min:   12.864 / max:  240.544) us
blackman            :    CPU:  164.734 us   +/-34.610 (min:  146.457 / max:  578.267) us     GPU:  168.007 us   +/-35.183 (min:  147.968 / max:  583.360) us

```
size = 100
```
blackman_fuse       :    CPU:   39.976 us   +/-13.275 (min:   34.264 / max:  400.114) us     GPU:   43.064 us   +/-13.989 (min:   36.768 / max:  405.568) us
blackman_PR         :    CPU:   13.138 us   +/- 6.808 (min:   10.804 / max:  182.403) us     GPU:   15.901 us   +/- 7.757 (min:   11.200 / max:  189.856) us
blackman            :    CPU:  162.628 us   +/-33.781 (min:  145.605 / max:  614.425) us     GPU:  165.866 us   +/-34.317 (min:  147.712 / max:  617.920) us

```
size = 1000
```
blackman_fuse       :    CPU:   41.210 us   +/-14.439 (min:   34.564 / max:  301.424) us     GPU:   44.361 us   +/-15.387 (min:   34.240 / max:  305.120) us
blackman_PR         :    CPU:   13.141 us   +/- 6.025 (min:   11.183 / max:  193.859) us     GPU:   15.770 us   +/- 6.740 (min:   11.296 / max:  196.992) us
blackman            :    CPU:  164.435 us   +/-33.677 (min:  147.302 / max:  546.201) us     GPU:  167.673 us   +/-34.191 (min:  150.240 / max:  549.280) us

```
size = 10000
```
blackman_fuse       :    CPU:   43.008 us   +/-15.630 (min:   35.168 / max:  195.289) us     GPU:   53.517 us   +/-15.960 (min:   43.328 / max:  203.712) us
blackman_PR         :    CPU:   14.469 us   +/- 7.382 (min:   11.064 / max:  255.856) us     GPU:   17.745 us   +/- 8.213 (min:   12.512 / max:  260.864) us
blackman            :    CPU:  156.967 us   +/-21.567 (min:  147.925 / max:  451.524) us     GPU:  160.348 us   +/-21.958 (min:  151.008 / max:  455.008) us

```
size = 100000
```
blackman_fuse       :    CPU:   38.669 us   +/- 9.138 (min:   35.625 / max:  221.387) us     GPU:  137.613 us   +/- 8.798 (min:  133.472 / max:  301.248) us
blackman_PR         :    CPU:   12.004 us   +/- 3.862 (min:   11.040 / max:  191.545) us     GPU:   26.858 us   +/- 3.986 (min:   24.000 / max:  194.624) us
blackman            :    CPU:  169.552 us   +/-40.674 (min:  150.714 / max:  526.531) us     GPU:  246.069 us   +/-29.993 (min:  230.624 / max:  548.064) us

```
size = 1000000
```
blackman_fuse       :    CPU:   48.030 us   +/-28.505 (min:   36.215 / max:  385.712) us     GPU: 1077.914 us   +/-21.551 (min: 1051.136 / max: 1368.448) us
blackman_PR         :    CPU:   14.551 us   +/- 8.022 (min:   11.099 / max:   96.597) us     GPU:  145.399 us   +/- 7.590 (min:  140.256 / max:  219.360) us
blackman            :    CPU:  184.252 us   +/-64.771 (min:  150.264 / max:  813.804) us     GPU: 2262.663 us   +/-28.294 (min: 2242.976 / max: 2537.856) us

```
using Elementwise kernels it is still significantly faster than fuse in this case.  

Ps. The last commit is for improvement i noticed when ran this benchmark, it seems that `cupy.creation.basic.ones(1, float)` is faster than `cupy.array([1])` and since that case is important I updated it to the previous implementation.

File: cupy/math/window.py [hunk 1]
```diff
@@ -2,10 +2,13 @@
...
-from cupy.creation import basic
-from cupy.creation import from_data
-from cupy.creation import ranges
-from cupy.math import trigonometric
+
+_blackman_kernel = core.ElementwiseKernel(
+    "float32 alpha",
+    "float64 out",
+    """
+    out = 0.42 - (0.5 * cos(i * alpha)) + (0.08 * cos(2 * alpha * i));
```
Comment (takagi) [inline cupy/math/window.py:10]: Would you remove extra parentheses?
```suggestion
    out = 0.42 - 0.5 * cos(i * alpha) + 0.08 * cos(2 * alpha * i);
```

File: cupy/math/window.py [hunk 2]
```diff
@@ -2,10 +2,13 @@
 
 import cupy
 from cupy import core
-from cupy.creation import basic
-from cupy.creation import from_data
-from cupy.creation import ranges
-from cupy.math import trigonometric
+
+_blackman_kernel = core.ElementwiseKernel(
```
Comment (takagi) [inline cupy/math/window.py:6]: Would you supply `name` parameter?

File: cupy/math/window.py [hunk 3]
```diff
@@ -28,13 +31,21 @@ def blackman(M):
...
     .. seealso:: :func:`numpy.blackman`
     """
-    if M < 1:
-        return from_data.array([])
     if M == 1:
-        return basic.ones(1, float)
-    n = ranges.arange(0, M)
-    return 0.42 - 0.5 * trigonometric.cos(2.0 * numpy.pi * n / (M - 1))\
-        + 0.08 * trigonometric.cos(4.0 * numpy.pi * n / (M - 1))
+        return cupy.creation.basic.ones(1, float)
```
Comment (takagi) [inline cupy/math/window.py:35]: Please use `cupy` module and dtype.
```suggestion
        return cupy.ones(1, dtype=cupy.float64)
```

File: cupy/math/window.py [hunk 4]
```diff
@@ -28,13 +31,21 @@ def blackman(M):
...
+    if M <= 0:
+        return cupy.creation.from_data.array([])
```
Comment (takagi) [inline cupy/math/window.py:37]: Please use `cupy` module.
```suggestion
        return cupy.array([])
```

File: cupy/math/window.py [hunk 5]
```diff
@@ -28,13 +31,21 @@ def blackman(M):
...
+    alpha = (numpy.pi * 2)/(M - 1)
```
Comment (takagi) [inline cupy/math/window.py:38]: nit: for style
```suggestion
    alpha = numpy.pi * 2 / (M - 1)
```

File: cupy/math/window.py [hunk 6]
```diff
@@ -28,13 +31,21 @@ def blackman(M):
...
+    out = cupy.empty(M, dtype=cupy.float64)
+    return _blackman_kernel(alpha, out)
+
+
+_hamming_kernel = core.ElementwiseKernel(
```
Comment (takagi) [inline cupy/math/window.py:43]: ditto (`name` parameter).

File: cupy/math/window.py [hunk 7]
```diff
@@ -56,12 +67,21 @@ def hamming(M):
 
     .. seealso:: :func:`numpy.hamming`
     """
-    if M < 1:
-        return from_data.array([])
     if M == 1:
-        return basic.ones(1, float)
-    n = ranges.arange(0, M)
-    return 0.54 - 0.46 * trigonometric.cos(2.0 * numpy.pi * n / (M - 1))
+        return cupy.creation.basic.ones(1, float)
```
Comment (takagi) [inline cupy/math/window.py:71]: ditto.
```suggestion
        return cupy.ones(1, dtype=cupy.float64)
```

File: cupy/math/window.py [hunk 8]
```diff
@@ -56,12 +67,21 @@ def hamming(M):
...
+    if M <= 0:
+        return cupy.creation.from_data.array([])
```
Comment (takagi) [inline cupy/math/window.py:73]: ditto.
```suggestion
        return cupy.array([])
```

File: cupy/math/window.py [hunk 9]
```diff
@@ -56,12 +67,21 @@ def hamming(M):
...
+    alpha = (numpy.pi * 2)/(M - 1)
```
Comment (takagi) [inline cupy/math/window.py:74]: ditto.
```suggestion
    alpha = numpy.pi * 2 / (M - 1)
```

File: cupy/math/window.py [hunk 10]
```diff
@@ -56,12 +67,21 @@ def hamming(M):
...
+    out = cupy.empty(M, dtype=cupy.float64)
+    return _hamming_kernel(alpha, out)
+
+
+_hanning_kernel = core.ElementwiseKernel(
```
Comment (takagi) [inline cupy/math/window.py:79]: ditto (`name` parameter).

File: cupy/math/window.py [hunk 11]
```diff
@@ -83,12 +103,13 @@ def hanning(M):
 
     .. seealso:: :func:`numpy.hanning`
     """
-    if M < 1:
-        return from_data.array([])
     if M == 1:
-        return basic.ones(1, float)
-    n = ranges.arange(0, M)
-    return 0.5 - 0.5 * trigonometric.cos(2.0 * numpy.pi * n / (M - 1))
+        return cupy.creation.basic.ones(1, float)
```
Comment (takagi) [inline cupy/math/window.py:107]: ditto.
```suggestion
        return cupy.ones(1, dtype=cupy.float64)
```

File: cupy/math/window.py [hunk 12]
```diff
@@ -83,12 +103,13 @@ def hanning(M):
...
+    if M <= 0:
+        return cupy.creation.from_data.array([])
```
Comment (takagi) [inline cupy/math/window.py:109]: ditto.
```suggestion
        return cupy.array([])
```

File: cupy/math/window.py [hunk 13]
```diff
@@ -83,12 +103,13 @@ def hanning(M):
...
+    alpha = (numpy.pi * 2)/(M - 1)
```
Comment (takagi) [inline cupy/math/window.py:110]: ditto.
```suggestion
    alpha = numpy.pi * 2 / (M - 1)
```

**Comment by niteya-shah** [comment]
@takagi I have made the changes.

**Comment by takagi** [comment]
LGTM, thanks!

**Comment by takagi** [comment]
pfnCI, test this please.

**Pull request merged**